File: agent-rag/api/conversations.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Request

from api.schemas import ChatHistoryResponse, ConversationsResponse, ConversationSummary, HistoryMessage
from db.repository import ConversationRepository
from db.session import get_db

router = APIRouter(prefix="/api/chat", tags=["chat-conversations"])


def _resolve_tenant_id(request: Request) -> str:
    return request.headers.get("X-Tenant-Id") or "default_tenant"


def _message_public_id(message) -> str:
    if message.client_message_id:
        return message.client_message_id
    return str(message.id)


@router.get("/conversations", response_model=ConversationsResponse)
def list_conversations(
    request: Request,
    limit: int = 50,
    offset: int = 0,
) -> ConversationsResponse:
    tenant_id = _resolve_tenant_id(request)
    limit = min(max(limit, 1), 100)
    offset = max(offset, 0)

    with get_db() as session:
        rows = ConversationRepository(session).list_conversations(tenant_id, limit=limit, offset=offset)
        summaries = [
            ConversationSummary(
                thread_id=row.thread_id,
                title=row.title,
                collection_name=row.collection_name,
                message_count=row.message_count,
                updated_at=row.updated_at,
            )
            for row in rows
        ]

    return ConversationsResponse(conversations=summaries)


@router.get("/history", response_model=ChatHistoryResponse)
def chat_history(thread_id: str, request: Request) -> ChatHistoryResponse:
    tenant_id = _resolve_tenant_id(request)

    with get_db() as session:
        conversation, messages = ConversationRepository(session).get_full_history(tenant_id, thread_id)
        if conversation is None:
            raise HTTPException(status_code=404, detail="thread not found")
        return ChatHistoryResponse(
            thread_id=thread_id,
            collection_name=conversation.collection_name,
            messages=[
                HistoryMessage(
                    id=_message_public_id(msg),
                    role=msg.role,
                    content=msg.content,
                    metadata=msg.metadata_json,
                )
                for msg in messages
            ],
        )


@router.delete("/conversations/{thread_id}")
def delete_conversation(thread_id: str, request: Request) -> dict[str, str]:
    tenant_id = _resolve_tenant_id(request)

    with get_db() as session:
        deleted = ConversationRepository(session).delete_conversation(tenant_id, thread_id)

    if not deleted:
        raise HTTPException(status_code=404, detail="thread not found")
    return {"status": "ok"}

File: agent-web/src/api/conversations.ts
import type { ConversationSummary, HistoryMessage } from '../types/chat'
import { buildHeaders, formatErrorWithRequestId, readRequestId } from './client'

const API_BASE = import.meta.env.VITE_API_BASE_URL ?? ''

export async function listConversations(
  tenantId: string,
  limit = 50,
  offset = 0,
): Promise<ConversationSummary[]> {
  const params = new URLSearchParams({ limit: String(limit), offset: String(offset) })
  const response = await fetch(`${API_BASE}/api/chat/conversations?${params}`, {
    headers: buildHeaders({ 'X-Tenant-Id': tenantId }),
  })
  if (!response.ok) {
    throw new Error(formatErrorWithRequestId(`加载会话列表失败 (${response.status})`, readRequestId(response)))
  }
  const data = (await response.json()) as { conversations: ConversationSummary[] }
  return data.conversations
}

export async function getChatHistory(tenantId: string, threadId: string): Promise<{
  thread_id: string
  collection_name: string | null
  messages: HistoryMessage[]
}> {
  const params = new URLSearchParams({ thread_id: threadId })
  const response = await fetch(`${API_BASE}/api/chat/history?${params}`, {
    headers: buildHeaders({ 'X-Tenant-Id': tenantId }),
  })
  if (response.status === 404) {
    throw new Error('会话不存在')
  }
  if (!response.ok) {
    throw new Error(formatErrorWithRequestId(`加载历史失败 (${response.status})`, readRequestId(response)))
  }
  return response.json()
}

export async function deleteConversation(tenantId: string, threadId: string): Promise<void> {
  const response = await fetch(`${API_BASE}/api/chat/conversations/${threadId}`, {
    method: 'DELETE',
    headers: buildHeaders({ 'X-Tenant-Id': tenantId }),
  })
  if (response.status === 404) {
    throw new Error('会话不存在')
  }
  if (!response.ok) {
    throw new Error(formatErrorWithRequestId(`删除会话失败 (${response.status})`, readRequestId(response)))
  }
}
